File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    queue = deque([0] * bridge_length)
    current_weight = 0 
    answer = 0
    
    truck_weights = truck_weights[::-1]  
    
    while truck_weights:
        current_weight -= queue.popleft()
        
        if current_weight + truck_weights[-1] <= weight:
            truck_weight = truck_weights.pop()
            queue.append(truck_weight)
            current_weight += truck_weight
        else:
            queue.append(0)
        
        answer += 1
    
    answer += bridge_length  
    
    return answer

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int total_weight = 0;
        Queue<Integer> queue = new LinkedList<>();
        int index = 0;
        while(index < truck_weights.length) {
            if(bridge_length == queue.size()) {
                total_weight -= queue.poll();
            }
            if(total_weight + truck_weights[index] <= weight) {
                total_weight += truck_weights[index];
                queue.offer(truck_weights[index]);
                index ++;
            } else {
                queue.offer(0);
            }
            answer += 1;
            
                
        }
        return answer + bridge_length;
    }
}
